### ISSUE TITLE
Custom friendlyClassName template

### ISSUE THREAD

**Issue by rofrischmann** [rofrischmann/react-look#241]
**Version**: 1.0.0
**Environment**: React
**Type**: Feature

-------------------
#### Description
Ability to add a custom `className` template to the `friendlyClassName`-plugin.
Right now it uses a default template which takes 3 parameters: `Component`, `element` and `className`. 
Using ES2015 it is quite easy to implement a custom template using template literals.

**Proposal (Code Example)** 
```javascript
const config = {
  // ... basic config
  friendlyClassNameTemplate: (className, Component, element) => {
     return Component.constructor.displayName + '-' + element.type + '__' + className
  }
}
```

**Proposal (How-To)** 
Instead of using the default template within the friendlyClassName-plugin you would first check for a custom template within the config (and only use the default one if not is provided).

### Want to try this? (first timers only!)
Get in contact and I can provide any help or further instructions on how this may be implemented.

**Comment by rofrischmann** [comment]
+1

**Comment by moaazsidat** [comment]
Hey Robin, would love to take this on and work on it.

**Comment by rofrischmann** [comment]
@moaazsidat Amazing :) Let me give some instructions tomorrow :) I am out this night (10pm here in Germany)

**Comment by moaazsidat** [comment]
No worries :)

**Comment by AlastairTaft** [comment]
Can we also expand the documentation with a little tutorial on how to get friendly class names working?

I've done what I believe to be the right thing but no luck getting the right output.

```javascript
import friendlyClassName from 'react-look/lib/plugins/friendlyClassName'
const serverConfig = Presets['react-dom']
serverConfig.plugins.push(friendlyClassName)
```

**Comment by rofrischmann** [comment]
@AlastairTaft Actually it seems to work for me with this configuration (while somehow it does not work always, but that's another issue which I might investigate later, there's no priority on devTools until the core is bulletproof :P) Still I recommend directly importing it from the exported Plugins
```javascript
import { Plugins, Presets } from 'react-look'

const config = Presets['react-dom']
config.plugins.push(Plugins.friendlyClassName)
```

@moaazsidat You should start with setting up the environment by following the Contributing.md guide. 
Done with that you should open *packages/react-look/modules/plugins/friendlyClassName.js* This plugin actually is kind of a mess and still I don't know how to do it better. 
Your focus should be line 20, `const newClass = classNameTemplate(Component, element, cls)`. Instead of always using the default template from line 4, we want to be able to pass a custom template with our configuration. Therefore we need to check if a `friendlyClassNameTemplate` was passed. You can get access to the config object via the plugin interface (just add `config` to the destructed parameters in line 10).

Now we got 2 cases:
1. custom template => call it with parameters `cls`, `Component` and `element` to get `newClass`
2. No custom template => use the default template as it is used right now

You should then add a new test case to verify the plugin works as expected. Open *packages/react-look/test/plugins/friendlyClassName-test.js* and add a new 
```javascript
it('should ...(sth. like: use a custom template)', () => { 
  /* test code here */
})
```
Basically just copy the tests above and add a mocked config object to the plugin interface with a random template and check the output.
Feel free to contact me via Gitter or Twitter if you need any further information or if my description actually is weird. I am not quite sure how to "describe" what needs to be done here.

**Comment by moaazsidat** [comment]
@rofrischmann, Thanks for the detailed explanation and outlining the steps of how to proceed. I'm wondering how the `friendlyClassNameTemplate` is passed within the `config` object. Is it a key, or is it a value belonging to another key in `config`?

Right now I have the below code starting on line 20:
```
const newClass;
if (config.friendlyClassNameTemplate) {
   newCLass = config.friendlyClassNameTemplate(cls, Component, element)  
} else {
   newClass = classNameTemplate(Component, element, cls)
}
```

**Comment by rofrischmann** [comment]
@moaazsidat Yeah, actually you can define it as how you'd like to, but basically I suggest `config.friendlyClassNameTemplate` is exactly what I thought of too. Note that you can not use `const` if you later assign a value, switch that to `let` to get it working :P 

I also just pushed 2 changes to correct the parameter order from `Component, element, cls` to `cls, Component, element`, you might want to check the latest commits. 
Then try to add the test and it's done :)

**Comment by moaazsidat** [comment]
I've added the following test: 
```
it('should use a custom template if one is provided in the config', () => {
    const styles = StyleSheet.create({
      box: {
        color: 'red',
        '@media (min-height: 300px)': {
          color: 'blue'
        }
      }
    })
    const Component = { constructor: { displayName: 'Comp' } }
    const element = <div></div>
    const newProps = { className: styles.box }
    const config = {
      friendlyClassNameTemplate: (cls, Component, element) => {
       return Component.constructor.displayName + '-custom-' + 'div' + '--custom--' + cls
      }
    }
    const pluginInterface = {
      StyleContainer,
      Component,
      config,
      element,
      newProps,
      styles: {}
    }

    friendlyClassName(pluginInterface)
    expect(newProps.className).to.eql('Comp-custom-div--custom--' + styles.box)

  })
```
However, it the test fails when I have the existing test run first (it succeeds when I comment out the old test). 

Result of running the test suite:
```
 184 passing (11s)
  1 failing

  1) Transforming classNames to friendly class names should use a custom template if one is provided in the config:

      AssertionError: expected 'Comp-div--c0' to deeply equal 'Comp-custom-div--custom--c0'
      + expected - actual

      -Comp-div--c0
      +Comp-custom-div--custom--c0
      
      at Context.<anonymous> (friendlyClassName-test.js:60:35)
```
Upon further inspection, I found that when the first test runs, and then the second one, the condition here is true, so the `else` isn't evaluated:
https://github.com/rofrischmann/react-look/blob/develop/packages/react-look/modules/plugins/friendlyClassName.js#L17-L21

Perhaps some side effect of the first test? I've added another `box` in my styles object in the test. I'm going to submit the PR in a bit, might be easier to explain on the PR.

**Comment by rofrischmann** [comment]
I see, the problem is that I got the tests setup to clean all static CSS styles before every test. This causes your new styles to be rendered as `c0` even if the first test already added this one to the classMapping. 
Not your fault but mine, I guess you should be able to test it by just adding another style to it, e.g. 
```javascript
const styles = StyleSheet.create({
  container: {},
  box: { /* styles here */ }
})
```
and use `styles.box` again. this case your `styles.box` will be rendered to `c1` and transformed correctly.

**Comment by moaazsidat** [comment]
Yeah, eventually figured that out :p 
I'll make the above changes to my PR to make it cleaner..
https://github.com/rofrischmann/react-look/pull/246/files

**Comment by rofrischmann** [comment]
Will be released today :)
